engine/execution/computation/programs_test.go

```diff
@@ -0,0 +1,368 @@
+}
+
+// TestPrograms_TestBlockForks tests the functionality of
+// programsCache under contract deployment and contract updates on
+// different block forks
+//
+// block structure and operations
+// Block1 (empty block)
+//     -> Block11 (deploy contract v1)
+//         -> Block111  (emit event - version should be 1) and (update contract to v3)
+//             -> Block1111   (emit event - version should be 3)
+//	       -> Block112 (emit event - version should be 1) and (update contract to v4)
+//             -> Block1121  (emit event - version should be 4)
+//     -> Block12 (deploy contract v2)
+//         -> Block121 (emit event - version should be 2)
+//             -> Block1211 (emit event - version should be 2)
+func TestPrograms_TestBlockForks(t *testing.T) {
+	// setup
+	rt := runtime.NewInterpreterRuntime()
+	chain := flow.Mainnet.Chain()
+	vm := fvm.New(rt)
+	execCtx := fvm.NewContext(zerolog.Nop(), fvm.WithChain(chain))
+
+	privateKeys, err := testutil.GenerateAccountPrivateKeys(1)
+	require.NoError(t, err)
+	ledger := testutil.RootBootstrappedLedger(vm, execCtx)
+	accounts, err := testutil.CreateAccounts(vm, ledger, fvm.NewEmptyPrograms(), privateKeys, chain)
+	require.NoError(t, err)
+
+	account := accounts[0]
+	privKey := privateKeys[0]
+
+	me := new(module.Local)
+	me.On("NodeID").Return(flow.ZeroID)
+
+	blockComputer, err := computer.NewBlockComputer(vm, execCtx, nil, nil, zerolog.Nop())
+	require.NoError(t, err)
+
+	programsCache, err := NewProgramsCache(10)
+	require.NoError(t, err)
+
+	engine := &Manager{
+		blockComputer: blockComputer,
+		me:            me,
+		programsCache: programsCache,
+	}
+
+	view := delta.NewView(ledger.Get)
+
+	var (
+		res *execution.ComputationResult
+
+		block1, block11, block111, block112, block1121,
+		block1111, block12, block121, block1211 *flow.Block
+
+		block1View, block11View, block111View, block112View, block1121View,
+		block1111View, block12View, block121View, block1211View *delta.View
+	)
+
+	t.Run("executing block1 (no collection)", func(t *testing.T) {
+		block1 = &flow.Block{
+			Header: &flow.Header{
+				View: 1,
+			},
+			Payload: &flow.Payload{
+				Guarantees: []*flow.CollectionGuarantee{},
+			},
+		}
+		block1View = view.NewChild()
+		executableBlock := &entity.ExecutableBlock{
+			Block: block1,
+		}
+		_, err := engine.ComputeBlock(context.Background(), executableBlock, block1View)
+		require.NoError(t, err)
+	})
+
+	t.Run("executing block11 (deploys contract version 1)", func(t *testing.T) {
+		block11tx1 := testutil.DeployEventContractTransaction(account, chain, 1)
+		prepareTx(t, block11tx1, account, privKey, 0, chain)
+
+		txs11 := []*flow.TransactionBody{block11tx1}
+		col11 := flow.Collection{Transactions: txs11}
+		block11View = block1View.NewChild()
+		block11, res = createTestBlockAndRun(t, engine, block1, col11, block11View)
+		// cache should include value for this block
+		require.NotNil(t, programsCache.Get(block11.ID()))
```

Comment:
Maybe its worth to add extra check with `HasChanges()`, just for sanity check. Essentially can be applied everywhere we asses programs

Reply:
added some cases